### ISSUE TITLE
ESC8 vulnerability is not detected

### ISSUE THREAD

**Issue by mashdk** [TrimarcJake/Locksmith#153]
Running Locksmith against a PKI, that is vulnerable to ESC8 (Web Enrollment installed, Extended Protection not enabled, NTLM enabled), does not raise an ESC8 issue.
Running PSPKIAudit against the same PKI, does raise an ESC8 issue

**Comment by techspence** [comment]
Because i'm a git noob and I'd end up screwing things up with a PR, I'll submit a POC here. Can you guys find a reason why this wouldn't be a valid way of detecting if the non-SSL version of these endpoints are up?

```PowerShell
$CADNSName = '<cadnsname>'
$CAName = '<caname>'

$OriginalErrorAction = $ErrorActionPreference
$ErrorActionPreference = 'SilentlyContinue'

foreach ($directory in @("foobar","certsrv/", "$($CAName)_CES_Kerberos/service.svc", "$($CAName)_CES_Kerberos/service.svc/CES", "ADPolicyProvider_CEP_Kerberos/service.svc", "certsrv/mscep/")) {
    $URL = "http://$CADNSName/$directory"
    $Request = Invoke-WebRequest -URI $URL -UseBasicParsing -UseDefaultCredentials
    if( -not $? )
    {
        $msg = $Error[0].Exception.Message
        if ($msg -match "401"){
            Write-Host "401 NOT AUTHORIZED BUT AVAILABLE: $URL"
        } elseif ($msg -match "404"){
            Write-Host "404 NOT FOUND: $URL"
        }
    } elseif ($Request.StatusCode -eq '200'){
        Write-Host "200 AVAILABLE AND AUTHORIZED: $URL"
    }
}
$ErrorActionPreference = $OriginalErrorAction
```

Example output...
![image](https://github.com/user-attachments/assets/6ed17b58-fd37-4ef1-82d1-1da225ae7313)

**Comment by TrimarcJake** [comment]
Updated to:
``` powershell
$CADNSName = '<cadnsname>'
$CAName = '<caname>'

$OriginalErrorAction = $ErrorActionPreference
$ErrorActionPreference = 'SilentlyContinue'

foreach ($directory in @("foobar","certsrv/", "$($CAName)_CES_Kerberos/service.svc", "$($CAName)_CES_Kerberos/service.svc/CES", "ADPolicyProvider_CEP_Kerberos/service.svc", "certsrv/mscep/")) {
    $URL = "http://$CADNSName/$directory"
    $Request = Invoke-WebRequest -URI $URL -UseBasicParsing -UseDefaultCredentials
    if( -not $? )
    {
        $msg = $Error[0].Exception.Message
        if ($msg -match "401"){
            Write-Host "401 NOT AUTHORIZED BUT AVAILABLE: $URL"
        } elseif ($msg -match "404"){
            Write-Host "404 NOT FOUND: $URL"
        } elseif ($msg -match '403') {
            Write-Host "403.4 HTTPS REQUIRED: $URL"
        }
    } elseif ($Request.StatusCode -eq '200'){
        Write-Host "200 AVAILABLE AND AUTHORIZED: $URL"
    }
}
$ErrorActionPreference = $OriginalErrorAction
```

Example output:
<img width="746" alt="image" src="https://github.com/user-attachments/assets/697e5bf7-4f7e-4d6b-89dc-0d823c70baea">

**Comment by TrimarcJake** [comment]
New version iterate through all endpoints in HTTP, then HTTPS + NTLM, then HTTPS + Negotiate and outputs as such:

```
Technique            : ESC8
Name                 : BrainstormCA
DistinguishedName    : CN=BrainstormCA,CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration,DC=BlueTuxedo,DC=DanglingSPNs,DC=lol
CAEnrollmentEndpoint : http://BTDC01.BlueTuxedo.DanglingSPNs.lol/certsrv/
AuthType             : NTLM
Issue                : An HTTP enrollment endpoint is available.
Fix                  : Disable HTTP access and enforce HTTPS.
                       Enable EPA.
                       Disable NTLM authentication (if possible.)

Technique            : ESC8
Name                 : BrainstormCA
DistinguishedName    : CN=BrainstormCA,CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration,DC=BlueTuxedo,DC=DanglingSPNs,DC=lol
CAEnrollmentEndpoint : https://BTDC01.BlueTuxedo.DanglingSPNs.lol/BrainstormCA_CES_Kerberos/service.svc
AuthType             : Negotiate
Issue                : An HTTPS enrollment endpoint is available.
Fix                  : Ensure EPA is enabled.
                       Disable NTLM authentication (if possible.)

Technique            : ESC8
Name                 : BrainstormCA
DistinguishedName    : CN=BrainstormCA,CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration,DC=BlueTuxedo,DC=DanglingSPNs,DC=lol
CAEnrollmentEndpoint : https://BTDC01.BlueTuxedo.DanglingSPNs.lol/ADPolicyProvider_CEP_Kerberos/service.svc
AuthType             : Negotiate
Issue                : An HTTPS enrollment endpoint is available.
Fix                  : Ensure EPA is enabled.
                       Disable NTLM authentication (if possible.)

Technique            : ESC8
Name                 : BrainstormCA
DistinguishedName    : CN=BrainstormCA,CN=Enrollment Services,CN=Public Key Services,CN=Services,CN=Configuration,DC=BlueTuxedo,DC=DanglingSPNs,DC=lol
CAEnrollmentEndpoint : http://BTDC01.BlueTuxedo.DanglingSPNs.lol/certsrv/mscep/
AuthType             : NTLM
Issue                : An HTTP enrollment endpoint is available.
Fix                  : Disable HTTP access and enforce HTTPS.
                       Enable EPA.
                       Disable NTLM authentication (if possible.)
```

**Comment by TrimarcJake** [comment]
https://github.com/TrimarcJake/Locksmith/tree/fix-esc8-false-negatives

**Comment by TrimarcJake** [comment]
Hey @mashdk ! I just added PS7 support to ESC8 detections. Can you test? https://github.com/TrimarcJake/Locksmith/tree/add-ps7-support-to-esc8-detections

**Comment by TrimarcJake** [comment]
Resolved in 2024.10.
